File: client_libs/js/encryption.js
const crypto = require('crypto');

/**
 * AES-GCM Encryption Utility (Node.js)
 * Compatible with Go implementation: AES-256, GCM, 12-byte Nonce prepended.
 */
class EncryptionUtil {
    static ALGORITHM = 'aes-256-gcm';
    static NONCE_LENGTH = 12;

    /**
     * Encrypts plaintext.
     * @param {string} keyBase64 - 32-byte key in Base64
     * @param {string} plaintext - Text to encrypt
     * @returns {string} Base64 encoded (Nonce + Ciphertext + Tag)
     */
    static encrypt(keyBase64, plaintext) {
        const key = Buffer.from(keyBase64, 'base64');
        if (key.length !== 32) throw new Error('Key must be 32 bytes (AES-256)');

        const nonce = crypto.randomBytes(this.NONCE_LENGTH);
        const cipher = crypto.createCipheriv(this.ALGORITHM, key, nonce);

        let ciphertext = cipher.update(plaintext, 'utf8');
        const final = cipher.final();
        const tag = cipher.getAuthTag();

        const combined = Buffer.concat([nonce, ciphertext, final, tag]);
        return combined.toString('base64');
    }

    /**
     * Decrypts ciphertext.
     * @param {string} keyBase64 - 32-byte key in Base64
     * @param {string} ciphertextBase64 - Base64 encoded (Nonce + Ciphertext + Tag)
     * @returns {string} Decrypted plaintext
     */
    static decrypt(keyBase64, ciphertextBase64) {
        const key = Buffer.from(keyBase64, 'base64');
        const input = Buffer.from(ciphertextBase64, 'base64');

        if (input.length < this.NONCE_LENGTH + 16) throw new Error('Ciphertext too short');

        const nonce = input.subarray(0, this.NONCE_LENGTH);
        const tag = input.subarray(input.length - 16); // Tag is last 16 bytes
        const ciphertext = input.subarray(this.NONCE_LENGTH, input.length - 16);

        const decipher = crypto.createDecipheriv(this.ALGORITHM, key, nonce);
        decipher.setAuthTag(tag);

        let plaintext = decipher.update(ciphertext, null, 'utf8');
        plaintext += decipher.final('utf8');

        return plaintext;
    }
}

module.exports = EncryptionUtil;

File: client_libs/dart/encryption.dart
import 'dart:convert';
import 'dart:typed_data';
import 'dart:math';

// Requires: encrypt package (pubspec.yaml: encrypt: ^5.0.0)
// This is a plain file, user must add dependency.
// For standard library only implementation, it's very hard in Dart without packages.
// Assuming 'encrypt' package is available or using pointycastle.
// I will provide code using 'encrypt' package as it is the standard.

// To verify: AES GCM Mode in 'encrypt' package.
// If not available easily, I will implement using PointyCastle logic if needed, 
// but sticking to high level 'encrypt' package is safer for users.
// Note: 'encrypt' package AES mode.gcm might be tricky.
// Let's use 'pointycastle' directly if we want pure compatibility without bulky wrappers?
// Actually 'encrypt' wraps pointycastle.

// Implementation using 'encrypt' package:
/*
import 'package:encrypt/encrypt.dart' as enc;

class EncryptionUtil {
  static String encrypt(String keyBase64, String plaintext) {
    final key = enc.Key.fromBase64(keyBase64);
    final iv = enc.IV.fromLength(12); // Random IV
    final encrypter = enc.Encrypter(enc.AES(key, mode: enc.AESMode.gcm));

    final encrypted = encrypter.encrypt(plaintext, iv: iv);
    
    // Output: Nonce (IV) + Ciphertext + Tag
    // Encrypted.bytes contains ciphertext + tag (usually).
    // Wait, encrypt package GCM keeps tag separate or inside?
    // It's safer to check documentation. 
  }
}
*/

// Since I cannot verify external packages easily, I will write this assuming the user adds `encrypt` package
// and handles the concatenation correctly. The `encrypt` package `Encrypted` object holds bytes. 
// GCM mode in `encrypt` puts tag at the end of bytes automatically in recent versions?
// Let's write a standard class that users can drop in, but warn about dependencies.

import 'package:encrypt/encrypt.dart';
import 'package:crypto/crypto.dart';

class EncryptionUtil {
  /// Encrypts plaintext using AES-GCM (256).
  /// Required pubspec: encrypt: ^5.0.3
  static String encrypt(String keyBase64, String plaintext) {
    final key = Key.fromBase64(keyBase64);
    
    // Generate 12 byte IV (Nonce)
    final iv = IV.fromSecureRandom(12);
    
    final encrypter = Encrypter(AES(key, mode: AESMode.gcm));
    
    final encrypted = encrypter.encrypt(plaintext, iv: iv);
    
    // encrypted.bytes usually includes the tag for GCM in this package.
    // Concatenate IV + EncryptedBytes
    final combined = Uint8List(iv.bytes.length + encrypted.bytes.length);
    combined.setAll(0, iv.bytes);
    combined.setAll(iv.bytes.length, encrypted.bytes);
    
    return base64.encode(combined);
  }

  /// Decrypts ciphertext.
  static String decrypt(String keyBase64, String ciphertextBase64) {
    final key = Key.fromBase64(keyBase64);
    final combined = base64.decode(ciphertextBase64);
    
    if (combined.length < 12 + 16) throw Exception("Ciphertext too short");
    
    // Extract IV (first 12 bytes)
    final iv = IV(combined.sublist(0, 12));
    
    // Extract Ciphertext+Tag (rest)
    final encryptedBytes = combined.sublist(12);
    final encrypted = Encrypted(encryptedBytes);
    
    final encrypter = Encrypter(AES(key, mode: AESMode.gcm));
    
    return encrypter.decrypt(encrypted, iv: iv);
  }
}
